capsules/src/net/sixlowpan.rs

```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
-    /// receive client per RxState did not make sense; thus, there is a single
-    /// receive client that receives all reassembled packets.
-    pub fn set_receive_client(&self, client: &'static ReceiveClient) {
-        self.rx_client.set(Some(client));
+    /// Sets the [SixlowpanClient](trait.SixlowpanClient.html) that will receive
+    /// transmission completion and new packet reception callbacks.
+    pub fn set_client(&'a self, client: &'a SixlowpanClient) {
+        self.client.set(Some(client));
     }
 
-    /// FragState::transmit_packet
-    /// --------------------------
-    /// This function is called to send a fully-formed IPv6 packet. Arguments
-    /// to this function are used to determine various aspects of the MAC
-    /// layer frame and keep track of the transmission state.
+    /// Transmits the supplied IPv6 packet.
+    ///
+    /// Transmitted IPv6 packets will be optionall compressed, fragmented and
+    ///
+    /// Only one transmission is allowed at a time. Calling this method while
+    /// before a previous tranismission has completed will return an error.
+    ///
+    /// # Arguments
+    ///
+    /// * `src_mac_addr` - Why is the argument specified?
+    ///
+    /// * `dst_mac_addr` - Why is the argument specified?
+    ///
+    /// * `ip6_packet` - A buffer containing the IPv6 packet to transmit. This
+    /// buffer will be passed back in the
+    /// [send_done](trait.SixlowpanClient.html#send_done) callback.
+    ///
+    /// * `ip6_packet_len` - The length of the packet, in case the `ip6_packet`
+    /// buffer is larger than the actual packet. Must be at most
+    /// `ip6_packet.len()`
+    ///
+    /// * `security` - An optional tuple specifying the security level
+    /// (encryption, MIC, or both) and key identifier to use. If elided, no
+    /// security is used.
+    ///
+    /// This function exposes the primary packet transmission fuctionality. The
+    /// source and destination mac address arguments specify the link-layer
+    /// addresses of the packet, while the `security`, `fragment`, and `compress`
+    /// options specify the security level (if any), whether to fragment the packet
+    /// if it is too large, and whether to compress the packet respectively. These
+    /// different options are exposed to provide the caller with more control over
+    /// how this layer modifies packets; no compression and no fragmentation can
+    /// be set, and the packet will be sent to the radio as a raw IPv6 packet.
+    ///
+    /// The `ip6_packet` argument contains a pointer to a buffer containing a valid
+    /// IPv6 packet, while the `ip6_packet_len` argument specifies the number of
+    /// bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
+    /// the invariant that `ip6_packet_len <= ip6_packet.len()`.
     pub fn transmit_packet(&self,
                            src_mac_addr: MacAddress,
```

Comment:
Why are is the transmitter specifying the source and destination Mac addresses? Or, _how_ does the caller of `transmit_packet` know the source and destination Mac addresses? Of course 6lowpan itself needs to know them but presumably they are determined by the Mac rather than, e.g., whoever is sending a UDP packet.

Reply:
So I'm not too sure what the final design will look like, but since the concept of gateways/routers are specific to Thread, I thought it was easiest to expose this functionality to the upper layers (since they will know our current gateway and hence the mac address of the next hop). The source address could be set and controlled by the lower layers however.

Reply:
Doesn't the Mac layer determine what our next hop is?

I think in this case it's better to underpromise the functionality at this stage and add it as necessary, but in any case if `dst_mac_addr` should be configurable by the client of `Sixlowpan` it should be with a setter method so we can at least set it during board configuration (or elsewhere) to some default. The client of `Sixlowpan` is, e.g., the UDP driver, which has no clue what a Mac address is

Reply:
So we could go with getters/setters, but this ties into another issue we're having; how and where to expose the configuration options for each layer. Should we allow some cross-layer configuration routines (e.g. UDP calls getters/setters for 6LoWPAN layer, bypassing the IP layer)?  But in this case, the upper layers do know what the Mac address is. Regardless, the `dst_mac_addr` option definitely needs to be changeable at runtime, as an end node must be able to change their gateway.

Also, if we ever want to support routing in the future, it may make sense to expose the destination mac address as an argument to send, as we would be dynamically dispatching/sending based on Thread-level state.